client/scala/src/main/scala/ai/verta/repository/Commit.scala

```diff
@@ -338,4 +338,78 @@ class Commit(
     else
       createCommit(message = message, diffs = diff.blobDiffs, commitBase = id)
   }
+
+  /** Generates a stream of outputs in this commit by walking through its folder tree.
+   *  The stream ends at the first failure, or when there are no folders left. If the commit is not saved, its only element is that Failure
+   *  @param walker the FolderWalker to process the folders in the folder tree.
+   *  @return a stream of Try's of WalkOutput.
+   */
+  def walk[T](walker: FolderWalker[T])(implicit ec: ExecutionContext): Stream[Try[T]] = {
+    if (!saved)
+      Stream(Failure(new IllegalCommitSavedStateException("Commit must be saved before it can be walked")))
+    else
+      continueWalk(getFolder(PathList(Nil)).get, walker)
+  }
+
+  /** Get the folder corresponding to a path in list form
+   *  @param location location
+   *  @return the folder, if succeeds
+   */
+  def getFolder(location: PathList)(implicit ec: ExecutionContext): Try[Folder] = {
+    clientSet.versioningService.GetCommitComponent2(
+      commit_sha = id.get,
+      repository_id_repo_id = repo.id,
+      location = if (location.components.length > 0) Some(location.components) else None
+    ).flatMap(r => {
+        val folderPath = location.path
+        val responseFolder = r.folder
+
+        val folderNames = responseFolder
+        .flatMap(_.sub_folders) // Option[List[VersioningFolderElement]]
+        .map(_.map(folder => folder.element_name.get))
```

Comment:
I believe this get might fail if you have no subfolders. I think the backend doesn't send the key instead of sending an empty list. Could you double check?

Reply:
Hmm responseFolder itself is an option, so I was trying to (flat)mapping it to avoid exception. Let me write a test on empty commit to see if it cause exception.

Reply:
Wrote a test on empty commit. See if it covers the case we're thinking of.

Reply:
Ok, maybe `element_name` can't actually be `None`. I checked this pretty late. Let me double check